lib/web_ui/lib/src/engine/compositor/skia_object_cache.dart

```diff
@@ -0,0 +1,238 @@
+      _itemMap.remove(oldObject);
+      oldObject.delete();
+    }
+  }
+}
+
+/// An object backed by a [js.JsObject] mapped onto a Skia C++ object in the
+/// WebAssembly heap.
+///
+/// These objects are automatically deleted when no longer used.
+abstract class SkiaObject {
+  /// The JavaScript object that's mapped onto a Skia C++ object in the WebAssembly heap.
+  js.JsObject? get skiaObject;
+
+  /// Deletes the associated C++ object from the WebAssembly heap.
+  void delete();
+}
+
+/// A [SkiaObject] that can resurrect its C++ counterpart.
+///
+/// Because there is no feedback from JavaScript's GC (no destructors or
+/// finalizers), we pessimistically delete the underlying C++ object before
+/// the Dart object is garbage-collected. The current algorithm deletes objects
+/// at the end of every frame. This allows reusing the C++ objects within the
+/// frame. In the future we may add smarter strategies that will allow us to
+/// reuse C++ objects across frames.
+///
+/// The lifecycle of a C++ object is as follows:
+///
+/// - Create default: when instantiating a C++ object for a Dart object for the
+///   first time, the C++ object is populated with default data (the defaults are
+///   defined by Flutter; Skia defaults are corrected if necessary). The
+///   default object is created by [createDefault].
+/// - Zero or more cycles of delete + resurrect: when a Dart object is reused
+///   after its C++ object is deleted we create a new C++ object populated with
+///   data from the current state of the Dart object. This is done using the
+///   [resurrect] method.
+/// - Final delete: if a Dart object is never reused, it is GC'd after its
+///   underlying C++ object is deleted. This is implemented by [SkiaObjects].
+abstract class ResurrectableSkiaObject extends SkiaObject {
+  ResurrectableSkiaObject() {
+    _skiaObject = createDefault();
+    SkiaObjects.manageResurrectable(this);
+  }
+
+  @override
+  js.JsObject? get skiaObject {
+    if (_skiaObject == null) {
+      _skiaObject = resurrect();
+      SkiaObjects.manageResurrectable(this);
+    }
+    return _skiaObject;
+  }
+
+  /// Do not use this field outside this class. Use [skiaObject] instead.
+  js.JsObject? _skiaObject;
+
+  /// Instantiates a new Skia-backed JavaScript object containing default
+  /// values.
+  ///
+  /// The object is expected to represent Flutter's defaults. If Skia uses
+  /// different defaults from those used by Flutter, this method is expected
+  /// initialize the object to Flutter's defaults.
+  js.JsObject createDefault();
+
+  /// Creates a new Skia-backed JavaScript object containing data representing
+  /// the current state of the Dart object.
+  js.JsObject resurrect();
+
+  @override
+  void delete() {
+    _skiaObject!.callMethod('delete');
+    _skiaObject = null;
+  }
+}
+
+/// A [SkiaObject] which is deleted once and cannot be used again.
+class OneShotSkiaObject extends SkiaObject {
```

Comment:
Everywhere `OneShotSkiaObject` is used, it's simply a wrapper around a `JsObject`, so since there was no special logic for resurrecting the `JsObject` it didn't need to be subclassed.